File: backend/shared/cache.ts
import redis from 'redis';
import * as dotenv from 'dotenv';

dotenv.config();

const redisHost = process.env.REDIS_HOST || 'localhost';
const redisPort = parseInt(process.env.REDIS_PORT || '6379', 10);
const redisPassword = process.env.REDIS_PASSWORD;

export const redisClient = redis.createClient({
  socket: {
    host: redisHost,
    port: redisPort,
  },
  password: redisPassword || undefined,
  legacyMode: false,
});

redisClient.on('error', (err) => {
  console.error('Redis Client Error', err);
});

redisClient.on('connect', () => {
  console.log('Redis client connected');
});

export async function getCache(key: string): Promise<any | null> {
  const value = await redisClient.get(key);
  if (value) {
    try {
      return JSON.parse(value);
    } catch {
      return value;
    }
  }
  return null;
}

export async function setCache(key: string, value: any, ttl: number = 3600): Promise<void> {
  const stringValue = typeof value === 'string' ? value : JSON.stringify(value);
  await redisClient.setEx(key, ttl, stringValue);
}

export async function deleteCache(key: string): Promise<void> {
  await redisClient.del(key);
}

export async function closeRedis(): Promise<void> {
  await redisClient.quit();
}

File: backend/api-service/src/cache.py
import redis.asyncio as redis
import json
from typing import Optional, Any
import os

REDIS_HOST = os.environ.get("REDIS_HOST", "redis")
REDIS_PORT = int(os.environ.get("REDIS_PORT", "6379"))
REDIS_DB = int(os.environ.get("REDIS_DB", "0"))
REDIS_PASSWORD = os.environ.get("REDIS_PASSWORD", "")

REDIS_URL = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/{REDIS_DB}" if REDIS_PASSWORD else f"redis://{REDIS_HOST}:{REDIS_PORT}/{REDIS_DB}"

_redis_client: Optional[redis.Redis] = None


async def get_redis() -> redis.Redis:
    global _redis_client
    if _redis_client is None:
        _redis_client = redis.from_url(REDIS_URL, decode_responses=True)
    return _redis_client


async def get_cache(key: str) -> Optional[Any]:
    client = await get_redis()
    value = await client.get(key)
    if value:
        return json.loads(value)
    return None


async def set_cache(key: str, value: Any, ttl: int = 300):
    client = await get_redis()
    await client.setex(key, ttl, json.dumps(value))


async def close_cache():
    global _redis_client
    if _redis_client:
        await _redis_client.close()
        _redis_client = None
